python-lib/dku_deeplearning_image/utils.py

```diff
@@ -77,25 +78,39 @@ def can_use_gpu():
     return len(tf.config.experimental.list_physical_devices('GPU')) > 0
 
 
-def set_gpu_options(should_use_gpu, gpu_list, memory_limit):
+def set_gpu_options(should_use_gpu, gpu_list, gpu_memory_allocation_mode, memory_limit_ratio=None):
     log_info("load_gpu_options")
     if should_use_gpu and can_use_gpu():
         log_info("should use GPU")
         gpus = tf.config.experimental.list_physical_devices('GPU')
+        for i, g in enumerate(gpus):
+            g.id = i
         gpus_to_use = [gpus[int(i)] for i in gpu_list] or gpus
-        if memory_limit:
+        log_info("GPUs on the machine: {}".format(str([g.id for g in GPUtil.getGPUs()])))
+        log_info("Will use the following GPUs: {}".format(gpus_to_use))
+        if gpu_memory_allocation_mode == constants.GPU_MEMORY_LIMIT and memory_limit_ratio:
             for gpu in gpus_to_use:
+                memory_limit = calculate_gpu_memory_allocation(memory_limit_ratio, gpu)
+                log_info("Restraining GPU(s) {} to {} Mo".format(gpu, str(memory_limit)))
                 tf.config.experimental.set_virtual_device_configuration(
                     gpu,
                     [tf.config.experimental.VirtualDeviceConfiguration(memory_limit=int(memory_limit))]
```

Comment:
I would advise against using anything from tf.config.experimental, but only tf.config.

Reply:
Mmmh this function typically is not available in the not experimental library. However, it is used here https://www.tensorflow.org/guide/gpu
I would rather not go on tf 2.4 because of Cuda versions

Reply:
Humm there are several CVE in tensorflow 2.2 which were fixed in later versions: https://github.com/tensorflow/tensorflow/releases

To me, that justifies the need to upgrade to 2.4, at the expense of older CUDA versions

Reply:
Will dedicate a new PR for that